gax/src/main/java/com/google/api/gax/tracing/ApiTracer.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.tracing;
+
+import com.google.api.core.BetaApi;
+import com.google.api.core.InternalExtensionOnly;
+import org.threeten.bp.Duration;
+
+/**
+ * Implementations of this class trace the logical flow of a google cloud client.
+ *
+ * <p>A single instance of a tracer represents a logical operation that can be annotated throughout
+ * its lifecycle. Constructing an instance of a subclass will implicitly signal the start of a new
+ * operation.
+ */
+@BetaApi("Surface for tracing is not yet stable")
+@InternalExtensionOnly
+public interface ApiTracer {
+  /**
+   * Asks the underlying implementation to install itself as a thread local. This allows for interop
+   * between clients using gax and external resources to share the same implementation of the
+   * tracing. For example OpenCensus will install a thread local that can read by the GRPC.
+   */
+  Scope inScope();
+
+  /**
+   * Signals that the overall operation has finished successfully. The tracer is now considered
+   * closed and should no longer be used.
+   */
+  void operationSucceeded();
+
+  /**
+   * Signals that the overall operation has failed and no further attempts will be made. The tracer
+   * is now considered closed and should no longer be used.
+   */
+  void operationFailed(Throwable error);
```

Comment:
Can you please add the documentation for arguments here and in other places? We do not add `@return` documentation element but usually add the `@param` (in some existing code it is missing, but that is not a pattern we should try to follow in new code).

Reply:
done